core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -408,11 +427,55 @@ private TableMetadata applyChangesToMapping(TableMetadata metadata) {
   private static Schema applyChanges(Schema schema, List<Integer> deletes,
                                      Map<Integer, Types.NestedField> updates,
                                      Multimap<Integer, Types.NestedField> adds,
-                                     Multimap<Integer, Move> moves) {
+                                     Multimap<Integer, Move> moves,
+                                     Set<String> identifierNames) {
+    // validate existing identifier fields are not deleted
+    for (String name : identifierNames) {
+      Types.NestedField field = schema.findField(name);
```

Comment:
If we have two columns `(a, b)` in the table, and the original identifier fields is `a`.  Then people want to delete the column `a` first and then rename the column `b` to column `a` and also change the identifier fields from old `a` to new `a`.  As the identifier is a constraint for delete, so people will need to call the API by : 

```java
SchemaUpdate update = ...
update.deleteColumn("a").renameColumn("b", "a")
```

In this way,  he will encounter an argument exception and try to call the `setIdentifierFields`: 

```java
update.setIdentifierFields("b")
  .deleteColumn("a")
  .renameColumn("b", "a")
  .setIdentiferFields("a")
```

But this way, people will still encounter the exception.  So maybe the final solution will be: 

```java
Step.1   update.setIdentifierFields(ImmutableSet.of()).commit();
Step.2   update.deleteColumn("a").renameColumn("b", "a").commit();
Step.3   update.setIdentifierFields("a").commit();
```

Reply:
I think multiple steps is right. We should not allow redefining the ID column's ID with a delete/rename.

Reply:
I listed the steps because I wanna to confirm that others are feeling good for this (I think it's OK). Seems like every one is OK, then I will mark this conversation resolved.